theano/compile/builders.py

```diff
@@ -168,30 +165,175 @@ def infer_shape(self, node, shapes):
 
         return ret
 
-    def grad(self, inputs, output_grads):
-        if hasattr(self, "grad_ops"):
-            grad_ops = self.grad_ops
-        else:
-            gs = theano.gradient.grad(cost=None,
-                                      known_grads=dict(izip(self.new_outputs,
-                                                            output_grads)),
-                                      wrt=self.new_inputs,
-                                      disconnected_inputs='ignore')
+    def perform(self, node, inputs, outputs):
+        raise NotImplementedError()
 
-            grad_ops = []
-            for g in gs:
-                if g is None:
-                    grad_ops.append(lambda *args: None)
-                else:
-                    # It is normal if some inputs are not needed in order
-                    # to compute the gradient, so we ignore them.
-                    grad_ops.append(OpFromGraph(self.new_inputs + output_grads,
-                                                [g],
-                                                on_unused_input='ignore'))
-            self.grad_ops = grad_ops
 
-        return [go(*(inputs + output_grads)) for go in grad_ops]
+class OpFromGraphPrecompiled(OpFromGraphBase):
+    """
+    The Op's inner graph is compiled into a theano function.
+    """
+    def prepare_node(self, node, storage_map, compute_map, impl):
+        if not hasattr(self, "fn") and impl == 'py':
+            self.fn = orig_function(self.internal_inputs,
+                                    self.internal_outputs,
+                                    **self.kwargs)
+
+    def perform(self, node, inputs, outputs):
+        variables = self.fn(*inputs)
+        assert len(variables) == len(outputs)
+        for output, variable in zip(outputs, variables):
+            # TODO: when function's output-borrowing semantics are correct,
+            # we wont need this copy anymore
+            output[0] = variable.copy()
+
+
+class OpFromGraphInline(OpFromGraphBase):
+    """
+    The Op's inner graph is expanded into the outer graph at compile time
+    """
+    def perform(self, node, inputs, outputs):
+        raise RuntimeError(
+            type(self).__name__ + ' is not supposed to be executed at runtime')
```

Comment:
Yes, a perform would be helpful for test values and DebugMode.
We had to do that for abstract ops, I think it is the case here as well.

What we could have (again, like for the abstract ops) is a pseudo-optimization in a final pass that asserts there is no Inline Ops in the graph. If we are compiling without any optimization at all, then that optimization will not be triggered and the slow version will be used.